turbopack/crates/turbo-tasks/src/scope.rs

```diff
@@ -82,6 +92,57 @@ impl ScopeInner {
             panic::resume_unwind(err);
         }
     }
+
+    fn worker(&self, first_job_index: usize, first_job: Box<dyn FnOnce() + Send + 'static>) {
+        let mut current_job_index = first_job_index;
+        let mut current_job = first_job;
+        loop {
+            let result = catch_unwind(AssertUnwindSafe(current_job));
+            let panic = result.err().map(|e| (e, current_job_index));
+            self.on_task_finished(panic);
+            let Some((index, job)) = self.pick_job_from_work_queue() else {
+                return;
+            };
+            current_job_index = index;
+            current_job = job;
+        }
+    }
+
+    fn pick_job_from_work_queue(&self) -> Option<(usize, Box<dyn FnOnce() + Send + 'static>)> {
+        let mut work_queue = self.work_queue.lock();
+        let job = loop {
+            if let Some(job) = work_queue.pop_front() {
+                break job;
+            } else {
+                self.work_queue_condition_var.wait(&mut work_queue);
+            };
+        };
+        match job {
+            WorkQueueJob::Job(index, job) => {
+                drop(work_queue);
+                Some((index, job))
+            }
+            WorkQueueJob::End => {
+                work_queue.push_front(WorkQueueJob::End);
```

Comment:
why do we re-push here?

leave a comment?

Reply:
ok i get it, still a comment here would be good to document the pattern

Reply:
I'll do a follow-up PR to make it more clear.

I think I refactor it to `struct WorkQueue { queue: VecDeque<...>, closed: bool }` this will avoid the `WorkQueueJob::End`